components/list/list-item-drag-mixin.js

```diff
@@ -33,86 +95,299 @@ export const ListItemDragMixin = superclass => class extends superclass {
+		this._destination = null;
+		this._visitedDestination = new Map();
+		this._moveBeforeDestination = false;
+		this._time = 0;
 	}
 
-	// TODO
-	_removeCopy() {
+	addVisitor(destination) {
+		if (destination && !this._visitedDestination.has(destination)) {
+			this._visitedDestination.set(destination, null);
+		}
+	}
 
+	clear() {
+		this._cleanUpOnLeave();
+		this._visitedDestination.forEach((_, visitedDestination) => visitedDestination._dropTarget = false);
+		this._visitedDestination.clear();
 	}
 
-	_renderDragAction(inner) {
-		return this.draggable ? html`
-			<div @click="${this._handleDragActionClick}" class="d2l-list-item-drag-action">${inner}</div>
-			` : nothing;
+	get destination() {
+		return this._destination;
 	}
-	_renderDragHandle() {
-		return this.draggable ? html`
-			<div
-				id="${this._itemDragId}"
-				class="d2l-list-item-drag"
-				@dragenter="${this._dragEnter}"
-				@dragleave="${this._dragExit}"
-			></div>
-		` : nothing;
+
+	get destinationKey() {
+		return this._destination && this._destination.key;
 	}
-};
 
-export class NewPositionEventDetails {
+	get moveBeforeDestination() {
+		return this._moveBeforeDestination;
+	}
+
+	setDestination(destination, moveBeforeDestination) {
+		this._moveBeforeDestination = moveBeforeDestination;
+		if (this._destination === destination) {
+			this._setPlacementMarkers();
+			return;
+		}
+		this._cleanUpOnLeave();
+		this._destination = destination;
+		this._setPlacementMarkers();
+		this.addVisitor(destination);
+	}
+
+	shouldDrop(time) {
+		return time - this._time < timeDelayForLeavingTheDropArea;
+	}
+
+	get target() {
+		return this._target;
+	}
+
+	get targetKey() {
+		return this._target && this._target.key;
+	}
+
+	updateTime(time) {
+		this._time = time;
+		if (this._timeoutId) clearTimeout(this._timeoutId);
+		this._timeoutId = setTimeout(() => {
+			this._cleanUpOnLeave();
+			this._destination = null;
+		}, timeDelayForLeavingTheDropArea);
+	}
+
+	_cleanUpOnLeave() {
+		if (!this._destination) return;
+		this._destination._topPlacementMarker = false;
+		this._destination._bottomPlacementMarker = false;
+		this._destination._inTopArea = false;
+		this._destination._inBottomArea = false;
+	}
+
+	_setPlacementMarkers() {
+		this._destination._topPlacementMarker = this.moveBeforeDestination;
+		this._destination._bottomPlacementMarker = !this.moveBeforeDestination;
+	}
+}
+
+class NewPositionEventDetails {
 	/**
 	 * @param { Object } object An simple object with the position event properties
 	 * @param { String } object.targetKey The item key of the list-item that is moving
 	 * @param { String } object.destinationKey The item key of the list-item in the position we are moving to
 	 * @param { String } object.temporaryMovement Information on whether the item is entering or exiting temporary movement
 	 */
-	constructor({targetKey, destinationKey, temporaryMovement}) {
+	constructor({targetKey, destinationKey, moveBeforeDestination, temporaryMovement}) {
```

Comment:
Can we add a comment to the function to specify this is a boolean parameter? It wasn't obvious to me at first

Reply:
Yes, also I am not a big fan of this name either. Any suggestions on something better?

Reply:
Maybe `insertBefore`?